server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java

```diff
@@ -122,27 +124,40 @@ public ReadOnlyEngine(EngineConfig config, SeqNoStats seqNoStats, TranslogStats
-        // Before 8.0 the global checkpoint is not known and up to date when the engine is created after
-        // peer recovery, so we only check the max seq no / global checkpoint coherency when the global
-        // checkpoint is different from the unassigned sequence number value.
-        // In addition to that we only execute the check if the index the engine belongs to has been
-        // created after the refactoring of the Close Index API and its TransportVerifyShardBeforeCloseAction
-        // that guarantee that all operations have been flushed to Lucene.
-        final Version indexVersionCreated = engineConfig.getIndexSettings().getIndexVersionCreated();
-        if (indexVersionCreated.onOrAfter(Version.V_7_1_0) ||
-            (seqNoStats.getGlobalCheckpoint() != SequenceNumbers.UNASSIGNED_SEQ_NO && indexVersionCreated.onOrAfter(Version.V_6_7_0))) {
-            if (seqNoStats.getMaxSeqNo() != seqNoStats.getGlobalCheckpoint()) {
-                throw new IllegalStateException("Maximum sequence number [" + seqNoStats.getMaxSeqNo()
-                    + "] from last commit does not match global checkpoint [" + seqNoStats.getGlobalCheckpoint() + "]");
+    /**
+     * Ensures that all operations exist in translog are committed into the given index commit. This check is mandatory
+     * so peer recovery of closed indices can skip phase 2 (i.e., not replaying translog operations) without losing data.
+     */
+    private void ensureNoUncommittedOperation(SeqNoStats seqNoStats, SegmentInfos segmentInfos) throws IOException {
+        // we can't enforce this check on an old index - should we prevent this engine as a recovery source?
+        if (config().getIndexSettings().getIndexVersionCreated().before(Version.V_6_7_0)) {
+            return;
+        }
+        final String translogUUID = segmentInfos.userData.get(Translog.TRANSLOG_UUID_KEY);
+        final Translog.TranslogGeneration recoverTranslogGeneration = new Translog.TranslogGeneration(
+            translogUUID, Long.parseLong(segmentInfos.userData.get(Translog.TRANSLOG_GENERATION_KEY)));
+        final TranslogDeletionPolicy translogDeletionPolicy = new TranslogDeletionPolicy(Long.MAX_VALUE, Long.MAX_VALUE);
+        translogDeletionPolicy.setTranslogGenerationOfLastCommit(recoverTranslogGeneration.translogFileGeneration);
+        translogDeletionPolicy.setMinTranslogGenerationForRecovery(recoverTranslogGeneration.translogFileGeneration);
+        final LocalCheckpointTracker localCheckpointTracker;
+        try (DirectoryReader reader = DirectoryReader.open(indexCommit)) {
+            localCheckpointTracker = createLocalCheckpointTracker(engineConfig, segmentInfos, logger,
+                () -> new Searcher("build_checkpoint_tracker", new IndexSearcher(reader), () -> {}), LocalCheckpointTracker::new);
+        }
+        try (Translog translog = new Translog(engineConfig.getTranslogConfig(), translogUUID, translogDeletionPolicy,
```

Comment:
I am a bit afraid of what would happen if that translog is very large? I mean we are reading the entire thing off disk no? Also isn't this expected to be empty or at least small?

Reply:
With https://github.com/elastic/elasticsearch/pull/41317/commits/8ba5ca5b70f8da88d6d5a8f29d4d4d7ec276ecd4,  we now only execute this check if we were a follower before and have gaps in sequence numbers. Even in that situation, we would read a few last translog generations since we read operations since the local checkpoint of the index commit.